[LeptonSFProducer] Rewrite the LeptonSFProducer. Less duplicates and easier to understand code.

cverstege opened the pull request:
Refactoring of the LeptonSFProducers, condensing common code.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
+        float sf2 = 1 / this->GetScaleFactor(0, *lep2);
+        product.m_weights["zl1" + weightName] = sf1;
+        product.m_weights["zl2" + weightName] = sf2;
+        product.m_weights["lepton" + weightName] = sf1 * sf2;
```
Sure you want to add these Weights to the `m_weights` map? When this producer is configured to run before the Weights-Producer, the produced event weight would be bogus. I would recommend adding only the one weight, which can be considered nominal and add the others to `m_optionalWeights`.

cverstege replied:
Updated

RHofsaess replied:
Wouldn't that still be a problem for the one weight?

cverstege replied:
What do you mean? Now only the "main" weight is in m_weights, all other weights, including those for individual muons are optional.

RHofsaess replied:
"When this producer is configured to run before the Weights-Producer, the produced event weight would be bogus" This, I mean.

cverstege replied:
No it would be correct, since only `product.m_weights["lepton" + weightName] = sf1 * sf2;` is set. All other SFs are stored in `m_optionalweights`. GitHub somehow didn't mark this code snippet as outdated. But it is, if you look at the last commit.

RHofsaess replied:
:+1:

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
+            product.m_weights["zl1" + weightName + "Up"] = sf1u;
+            product.m_weights["zl1" + weightName + "Down"] = sf1d;
+            product.m_weights["zl2" + weightName + "Up"] = sf2u;
+            product.m_weights["zl2" + weightName + "Down"] = sf2d;
```
Same here. These weights don't belong to the nominal weight map.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -248,77 +233,12 @@ std::string LeptonTrackingSFProducer::GetProducerId() const { return "LeptonTrac
-    delete sfhisto;
-    file.Close();*/
-}
+void LeptonTrackingSFProducer::Init(ZJetSettings const& settings) {}
```
Is this empty on purpose?

HerrHorizontal replied:
In case this should not be provided, add some debug output or throw an error.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -248,77 +233,12 @@ std::string LeptonTrackingSFProducer::GetProducerId() const { return "LeptonTrac
-            product.m_weights["leptonTrackingSFWeightDown"] = 0;
-        }
-    }*/
 }
```
Same here. It seems there is something missing.

cverstege replied:
Before it was empty as well, I removed the old commented code. Currently no tracking SFs are provided. I could add a log fatal.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -329,148 +249,64 @@ std::string LeptonTriggerSFProducer::GetProducerId() const { return "LeptonTrigg
+        double sf2 = 1 / this->GetScaleFactor(0, *lep2);
+        product.m_weights["zl1" + weightName] = sf1;
+        product.m_weights["zl2" + weightName] = sf2;
+        product.m_weights["lepton" + weightName] =
```
Same as above. Add them to different weight object.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -329,148 +249,64 @@ std::string LeptonTriggerSFProducer::GetProducerId() const { return "LeptonTrigg
+                             (1 - GetScaleFactor(1, *lep2) + 0.005 * GetScaleFactor(0, *lep2)));
+            product.m_weights["lepton" + weightName + "Down"] =
+                1 / (1 - (1 - GetScaleFactor(-1, *lep1) - 0.005 * GetScaleFactor(0, *lep1)) *
+                             (1 - GetScaleFactor(-1, *lep2) - 0.005 * GetScaleFactor(0, *lep2)));
```
Same here.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -329,148 +249,64 @@ std::string LeptonTriggerSFProducer::GetProducerId() const { return "LeptonTrigg
+            if (settings.GetDebugVerbosity() > 1) {
+                LOG(DEBUG) << "Lepton1 SF up: " << sf1u << ", Lepton1 SF down: " << sf1d;
+                LOG(DEBUG) << "Lepton2 SF up: " << sf2u << ", Lepton2 SF down: " << sf2d;
+                LOG(DEBUG) << "Total SF up: " << product.m_weights.at("lepton" + weightName + "Up")
```
You could think to temporarily save the value inside this scope to avoid searching for the value in the map again. This might save some unnecessary instructions for the price of temporary acquiring a few additional bytes in memory.

HerrHorizontal reviewed (changes requested):
This looks way more elegant now. Thank you very much. I have only some minor proposals for improvement (see inline comments).

HerrHorizontal reviewed (approved):
This looks fine now.

RHofsaess commented on Compile/interface/Producers/LeptonSFProducer.h:
```diff
@@ -16,192 +18,75 @@ class LeptonSFProducer : public ZJetProducerBase
-    bool m_etarev2D[3] = {};
+    std::string weightName;
     bool m_reversed_axes;
     bool m_absoluteeta;
```
I know, we don't have a consistent naming convention and this was already beforem but may change the variable name to `m_absoluteEta`?

cverstege replied:
Ok

RHofsaess commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
-    
+    // Get file
+    LOG(INFO) << this->GetProducerId() << ": Loading file " << m_sffile << ", Histogram "
+              << histoname;
```
Is histoname here already set?

cverstege replied:
This is set by the individual producers, which call the Parent Init at the end.

RHofsaess replied:
okay, I wasn't sure if this producer could run standalone.
Then it is okay.

RHofsaess commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
+            LOG(INFO) << this->GetProducerId() << ": eta-pt map used for SF";
+        } else if (histoname.find("pt_eta") != std::string::npos) {
+            m_reversed_axes = true;
+            LOG(INFO) << this->GetProducerId() << ": pt-eta map (reversed axes) used for SF";
```
.
:D

RHofsaess commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
+{
+    int nEta = m_reversed_axes ? sfhisto->GetNbinsY() : sfhisto->GetNbinsX();
+    int nPt = m_reversed_axes ? sfhisto->GetNbinsX() : sfhisto->GetNbinsY();
+    for (int i = 1; i <= nEta; ++i) {
```
Maybe add a comment for clarity, why it starts with 1 (underflow).

cverstege replied:
Ok

RHofsaess commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
+            }
+        }
+    } else {
+        LOG(DEBUG) << "No valid Z found. Only applying SFs to Z leptons.";
```
What is ment with "Z leptons" in this case?
Since no valid Z is available?

cverstege replied:
Yes, I only apply the SFs to the leptons from a Valid Z. Applying them to a third valid Muon is wrong, as we only care about the two Z muons.

RHofsaess replied:
But if m_validZ is false, we have no "Z leptons". Therefore, I find this Debug message missleading.

cverstege replied:
True, and then No SFs are calculated and stored, which is what I want. The default value for weights, if they are not set is 1. Which should be fine.

RHofsaess replied:
Ok, so this actually means that nothing is applied. I was a little bit confused.

RHofsaess commented:
Very nice, thanks!
I only left some minor inline comments.

cverstege commented:
Some changes were made. I reverted back to using optionalWeights, as this makes things easier when comapring Reco lovevl with efficiencies to Gen level. The main weight for gen level should not be changed by the efficiencies at reco level.
Additionally I implemented "correct" uncertainty handling according to the MuonPOG: https://cms-talk.web.cern.ch/t/questions-on-the-combination-of-muon-tnp-uncertainties/5968
Please re-review @HerrHorizontal and @RHofsaess

cverstege commented:
I'll squash and merge one approved, so it's only one commit on the master branch in the end.

HerrHorizontal commented on Compile/interface/Producers/LeptonSFProducer.h:
```diff
@@ -203,5 +87,6 @@ class LeptonRecoSFProducer : public LeptonSFProducer{
 
   protected:
-    virtual float GetScaleFactor(float err_shift, KLV const& lepton) const;
+    virtual std::tuple<float, float> GetScaleFactorAndUnc(KLV const& lepton) const;
```
You should also add an `overwrite` here, since this method is virtual in the parent class.

HerrHorizontal replied:
Also you shouldn't declare it as `virtual`, since this will not overwritten in any child.

HerrHorizontal commented on Compile/interface/Producers/LeptonSFProducer.h:
```diff
@@ -203,5 +87,6 @@ class LeptonRecoSFProducer : public LeptonSFProducer{
   protected:
-    virtual float GetScaleFactor(float err_shift, KLV const& lepton) const;
+    virtual std::tuple<float, float> GetScaleFactorAndUnc(KLV const& lepton) const;
+    virtual void SetOverflowPtBin();
```
Also here: no `virtual` but a `overwrite` instead.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -54,92 +178,25 @@ std::string LeptonIDSFProducer::GetProducerId() const { return "LeptonIDSFProduc
-        product.m_weights["leptonIDSFWeightUp"] = 0;
-        product.m_weights["leptonIDSFWeightDown"] = 0;
-    }
+    LeptonSFProducer::Produce(event, product, settings);
```
If you are not doing anything different in the `Produce` method of the parent class. Why do you need to define it here and declare it as `overwrite` in the header?

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -150,94 +207,25 @@ std::string LeptonIsoSFProducer::GetProducerId() const { return "LeptonIsoSFProd
-            product.m_weights["leptonIsoSFWeightDown"] = 0;
-        }
-    }
+    LeptonSFProducer::Produce(event, product, settings);
```
Same here.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -482,222 +319,194 @@ std::string LeptonRecoSFProducer::GetProducerId() const { return "LeptonRecoSFPr
-    } else {
-        LOG(DEBUG) << "No valid Z found. Only applying SFs to Z leptons.";
-    }
+    LeptonSFProducer::Produce(event, product, settings);
```
Also here.

HerrHorizontal commented on Compile/src/Producers/LeptonSFProducer.cc:
```diff
@@ -482,222 +319,194 @@ std::string LeptonRecoSFProducer::GetProducerId() const { return "LeptonRecoSFPr
+    }
+}
+
+std::tuple<float, float> LeptonRecoSFProducer::GetHighPtEff(KLV const& lepton) const
```
Is this method declared in the header?

cverstege replied:
Yes

HerrHorizontal reviewed (changes requested):
I just stumbled upon some points, which need some clearance.
